Implement bulk polling of volumes

gnufied opened the pull request:
This implements Bulk volume polling using ideas presented by
justin in https://github.com/kubernetes/kubernetes/pull/39564

But it changes the implementation to use an interface
and doesn't affect other implementations.

cc @justinsb

gnufied committed:
Impement bulk polling of volumes

This implements Bulk volume polling using ideas presented by
justin in https://github.com/kubernetes/kubernetes/pull/39564

But it changes the implementation to use an interface
and doesn't affect other implementations.

k8s-reviewable commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.kubernetes.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/41306)
<!-- Reviewable:end -->

jingxu97 commented on pkg/volume/util/operationexecutor/operation_executor.go:
```diff
@@ -387,6 +391,79 @@ func (oe *operationExecutor) DetachVolume(
+	if err != nil {
+		return err
+	}
+	return oe.pendingOperations.Run("cluster-bulk-volume-poll", "" /* podName */, bulkVerifyVolumeFunc)
```
"cluster-bulk-volume-poll" this is a fixed string. So it will prevent any concurrent VerifyVolumeFunc goroutine no matter what type of volume plugins?

gnufied replied:
okay - yeah I am going to change this to per-plugin basis.

justinsb commented:
So I actually do like this, as well as the caching layer :-).  And the interface approach is much cleaner.  I'm going to dig in to the code to understand the following two things I didn't understand:

1) Why we can't delete the non-bulk code
2) Whether there is value in returning the node a volume is actually attached to

justinsb commented on pkg/volume/aws_ebs/attacher.go:
```diff
@@ -248,3 +293,18 @@ func (detacher *awsElasticBlockStoreDetacher) Detach(deviceMountPath string, nod
+	nodeName types.NodeName,
+	check bool) {
+
+	volumeMap, volumeMapExists := nodeDiskMap[nodeName]
```
For a `map[]map...`, we can probably write this more traditionally:

```
volumeMap := nodeDiskMap[nodeName]
if volumeMap == nil {
  volumeMap = make(....)
  nodeDiskMap[nodeName] = volumeMap
}
volumeMap[volumeSpec] = check
```

justinsb commented on pkg/volume/util/operationexecutor/operation_executor.go:
```diff
@@ -387,6 +391,79 @@ func (oe *operationExecutor) DetachVolume(
+				continue
+			}
+
+			volumeAttacher, newAttacherErr := attachableVolumePlugin.NewAttacher()
```
So this is a little different from the code before - it creates a new attacher for every volume AFAICT.  The old code did one attacher per plugin name

gnufied replied:
Yeah but this one only exists to verify in plugin really supports Bulk verification, I am planning to refactor this one bit here.

justinsb commented on pkg/volume/util/operationexecutor/operation_executor.go:
```diff
@@ -387,6 +391,79 @@ func (oe *operationExecutor) DetachVolume(
+				continue
+			}
+			// If node doesn't support Bulk volume polling it is best to poll individually
+			oe.VerifyVolumesAreAttached(nodeAttachedVolumes, node, actualStateOfWorld)
```
I think this is calling once per volume, instead of once per node here?

justinsb replied:
Also we're not checking the error code here.

gnufied replied:
This is still calling once per node, not per volume because `nodeAttachedVolumes` is slice of all volumes attached to the node. Agreed about error code though.

justinsb commented on pkg/volume/util/operationexecutor/operation_executor.go:
```diff
@@ -387,6 +391,79 @@ func (oe *operationExecutor) DetachVolume(
+			break
+		}
+	}
+	bulkVerifyVolumeFunc, err :=
```
(We could skip this if bulkVerifyPluginsByNode, but it seems harmless)

gnufied replied:
changed this whole thing

justinsb commented on pkg/volume/util/operationexecutor/operation_generator.go:
```diff
@@ -167,6 +174,66 @@ func (og *operationGenerator) GenerateVolumesAreAttachedFunc(
+					nodeVolumeSpecs, nodeChecked := attached[nodeName]
+
+					if !nodeChecked {
+						actualStateOfWorld.MarkVolumeAsDetached(volumeSpecMap[volumeSpec], nodeName)
```
missing `continue`?

I think you could also start with `check := false`, and then update it `if nodeVolumeSpecs != nil`

justinsb commented on pkg/volume/util/operationexecutor/operation_generator.go:
```diff
@@ -167,6 +174,66 @@ func (og *operationGenerator) GenerateVolumesAreAttachedFunc(
 	}, nil
 }
 
+func (og *operationGenerator) GenerateBulkVolumeVerifyFunc(
```
I would be so much happier if we didn't have two parallel functions, though I realize that is tricky.

One option _could_ be to return an `[]Operation`, which would include the key & the function.

saad-ali replied:
+1

gnufied replied:
I am going to create some issues and if try and get saad/jing to assign them to myself.

gnufied commented on pkg/cloudprovider/providers/aws/aws.go:
```diff
@@ -1786,6 +1803,61 @@ func (c *Cloud) DisksAreAttached(diskNames []KubernetesVolumeID, nodeName types.
 	return attached, nil
 }
```
I am half thinking that may be - we should delete the whole `DisksAreAttached` and `VolumesAreAttached` functions and make `VolumesAreAttached` function for aws log a giant warning/error when it gets used here. Because really no one should be using it.

gnufied commented on pkg/volume/aws_ebs/attacher.go:
```diff
@@ -112,6 +112,51 @@ func (attacher *awsElasticBlockStoreAttacher) VolumesAreAttached(specs []*volume
 	return volumesAttachedCheck, nil
 }
```
I got rid of `DisksAreAttached` function.

gnufied commented:
@justinsb  @jingxu97 I have fixed most of the review comments. Here is a summary of changes:

1. Enable per-plugin bulk volume operation generator so that different volume plugins can be bulk polled concurrently. In real world though, I don't think this is going to have major impact since most attach/detach plugins we have require cloudprovider configuration and I don't think currently kubernetes supports such hybrid configuration of running with multiple cloud provider configurations. But it is still good change because rather than hardcoding some random string I can use plugin name as key for `Run` method. 

2. Cleaned up and got rid of `DisksAreAttached` functions.

3. Cleaned up the way `BulkVerifyVolumes` gets initialized in the `operation_executator.go`. I think the new change is cleaner since I moved check for BulkVolume support from attacher interface to `VolumePlugin` interface.

jingxu97 commented on pkg/cloudprovider/providers/aws/aws.go:
```diff
@@ -1777,36 +1791,55 @@ func (c *Cloud) DiskIsAttached(diskName KubernetesVolumeID, nodeName types.NodeN
 	}
-	_, instance, err := c.getFullInstance(nodeName)
+
+	awsInstances, err := c.getAllRunningInstances()
```
why not filter only the nodes in the map? We don't need to getAllRunningInstances here.

gnufied replied:
fixed.

jingxu97 commented on pkg/volume/volume.go:
```diff
@@ -187,6 +187,13 @@ type Attacher interface {
 }
 
+type BulkVolumeVerifier interface {
+	// VolumesAreAttached checks whether the list of volumes still attached to the specified
```
modify the comment to match BulkVerifyVolumes

gnufied replied:
fixed.

jingxu97 commented on pkg/cloudprovider/providers/aws/aws.go:
```diff
@@ -330,8 +330,7 @@ type Volumes interface {
 
-	// Check if a list of volumes are attached to the node with the specified NodeName
-	DisksAreAttached(diskNames []KubernetesVolumeID, nodeName types.NodeName) (map[KubernetesVolumeID]bool, error)
+	CheckClusterDisks(map[types.NodeName][]KubernetesVolumeID) (map[types.NodeName]map[KubernetesVolumeID]bool, error)
```
why not still use DisksAreAttached? I think this makes aws API different from other cloud provider?

gnufied replied:
`DisksAreAttached` is internal implementation detail, it just happens to be used in other cloud providers too because people copied some of the ideas of implementation. I don't think `DisksAreAttached` is part of `VolumePlugin` interface or anything.

jingxu97 replied:
put comments for this API. Just to be consistent, could we still use DisksAreAttached? Any reason you don't want to use the same name as others?

jingxu97 replied:
Ping on this

gnufied replied:
hmm, so just renaming the function or event putting it back doesn't get rid of change from admission controller. if if use `DisksAreAttached` same as `CheckClusterDisks` then function signature is changed and that would still mean changing admission plugin tests.

gnufied replied:
I have renamed `CheckClusterDisks` to `DisksAreAttached`

justinsb commented on pkg/cloudprovider/providers/aws/aws.go:
```diff
@@ -3201,7 +3239,20 @@ func (c *Cloud) getInstancesByNodeNamesCached(nodeNames sets.String) ([]*ec2.Ins
 	}
-	names := aws.StringSlice(nodeNames.List())
+	instances, err := c.getInstancesByNodeNames(nodeNames.List())
+
```
Nit: Please don't leave a line in between the function and the `if err != nil`

gnufied replied:
fixed

jingxu97 commented on pkg/cloudprovider/providers/aws/aws.go:
```diff
@@ -1066,6 +1065,21 @@ func (c *Cloud) getInstancesByRegex(regex string) ([]types.NodeName, error) {
 }
 
+// getAllRunningInstances returns all running instances that are part of this cluster
+func (c *Cloud) getAllRunningInstances() ([]*ec2.Instance, error) {
```
this function is not used?

gnufied replied:
ack

gnufied replied:
ack

justinsb commented on pkg/cloudprovider/providers/aws/aws.go:
```diff
@@ -3201,7 +3239,20 @@ func (c *Cloud) getInstancesByNodeNamesCached(nodeNames sets.String) ([]*ec2.Ins
+	instances, err := c.getInstancesByNodeNames(nodeNames.List())
+
+	if err != nil {
+		return nil, err
```
I think you've changed the behaviour here for nil, nil

gnufied replied:
Do you mean `getInstancesByNodeNames` returns `nil,nil` when no instances are found?

gnufied replied:
fixed

justinsb commented on pkg/cloudprovider/providers/aws/aws.go:
```diff
@@ -3314,3 +3361,18 @@ func (c *Cloud) addFilters(filters []*ec2.Filter) []*ec2.Filter {
 	return c.filterTags[TagNameKubernetesCluster]
 }
+
+func setNodeDisk(
```
Let's inline this

gnufied replied:
We are using this function in 2 place @justinsb , so if we inline this we will have to copy same 3-4 lines in 2 places. perhaps lets leave it like how it is?

jingxu97 replied:
maybe put some comments for this new function or use a more detailed function name?

justinsb reviewed (commented):
Some comments on the AWS side of things

jingxu97 commented on pkg/cloudprovider/providers/aws/aws.go:
```diff
@@ -1777,36 +1791,60 @@ func (c *Cloud) DiskIsAttached(diskName KubernetesVolumeID, nodeName types.NodeN
-			// Disk is still attached to node
-			attached[diskName] = true
+		idToDiskName := make(map[awsVolumeID]KubernetesVolumeID)
+		for _, diskName := range diskNames {
```
Can we combine this loop with the previous one in line 1817?

jingxu97 replied:
ping on this?

gnufied replied:
agreed. I will modify.

gnufied replied:
it is bit tricky to roll this into same loop. We want by default disks to be marked as `detached`, that is what loop at line#1817 does. But if the node has disappeared from AWS, the we do `continue` and move to next node. But all disks on that node are considered detached.

If we combine the loops, then we wouldn't be able to bail out to next `continue` if node doesn't exist in AWS. 

It is of course possible to rearrange the code so that, we don't depend on the continue line:

```
		awsInstance := awsInstanceMap[nodeName]
		if awsInstance == nil {
			// If instance no longer exists, safe to assume volume is not attached.
			glog.Warningf(
				"Node %q does not exist. DisksAreAttached will assume disks %v are not attached to it.",
				nodeName,
				diskNames)
			continue
		}
```

But I feel we might be okay. @jingxu97 what you think?

gnufied replied:
I have moved the first loop to line #1787. PTAL

jingxu97 replied:
got it. thanks!

jingxu97 commented on pkg/volume/util/operationexecutor/operation_executor.go:
```diff
@@ -387,6 +391,82 @@ func (oe *operationExecutor) DetachVolume(
+			}
+
+			// If node doesn't support Bulk volume polling it is best to poll individually
+			nodeError := oe.VerifyVolumesAreAttached(nodeAttachedVolumes, node, actualStateOfWorld)
```
if a node has multiple volume plugins, some of them not support bulkvolumepolling, and some support. Will this logic has issue?

gnufied replied:
Currently we don't have this problem because attachable volume types also require cloudprovider configuration, in future in worst case we would be individually polling all such plugins on node and won't be using bulk polling. I don't think this can can be helped though.

jingxu97 replied:
This function is much larger than other functions in this file. The logic is a little bit tricky, likethe "break" is very critical in line 449. Could we change something like following?
bulkVerifyVolumesByPlugin := make(map[string]map[types.NodeName][]AttachedVolume)
nonBulkVerifyVolumesByPlugin := make(map[string]map[types.NodeName][]AttachedVolume)
for node, nodeAttachedVolumes := range attachedVolumes {
 		for _, volumeAttached := range nodeAttachedVolumes {
 			// get volumePlugin
                        // If plugin support bulkVolume, put in bulkVerifyVolumesByPlugin, 
                        // otherwise put in nonBulkVerifyVolumesByPlugin

 for pluginName, bulkVerifyVolumes := range bulkVerifyVolumesByPlugin
             // oe.operationtor.GenerateBulkVolumeBerifyFunc( bulkVerifyVolumes, actualStateOfWorld)

for pluginName, nonBulkVerifyVolumes := range nonBulkVerifyVolumesByPlugin
           for nodeName, volumes := range nonBulkVerifyVolumes 
           //oe.VerifyVolumesAttached(volumes, nodeName, actualStateOfWorld)

This might involves more looping, but I feel the logic is easier and code change is smaller.

jingxu97 commented on pkg/controller/volume/attachdetach/reconciler/reconciler.go:
```diff
@@ -121,12 +121,7 @@ func (rc *reconciler) updateSyncTime() {
-			glog.Errorf("Error in syncing states for volumes: %v", err)
-		}
-	}
+	rc.attacherDetacher.BulkVerifyVolumes(volumesPerNode, rc.actualStateOfWorld)
```
Is that ok we still use VerifyVolumesAreAttached instead of changing to BulkVerifyVolume. For some plugins, it might not be bulk polling. Also this name is used in multiple places which is a little hard to differentiate.

gnufied replied:
hmm, not bad idea. I can change it.

gnufied replied:
fixed.

jingxu97 commented on pkg/cloudprovider/providers/aws/aws.go:
```diff
@@ -1777,36 +1776,64 @@ func (c *Cloud) DiskIsAttached(diskName KubernetesVolumeID, nodeName types.NodeN
+
+	if len(awsInstances) == 0 {
+		glog.V(3).Infof("No instances found with volumes still attached to node")
+		return nil, nil
```
if no instances returned, should we assume that all volumes are not attached like the l logic below?

gnufied replied:
ack. I was planning to return `attached` here

gnufied replied:
fixed.

jingxu97 commented on pkg/volume/util/operationexecutor/operation_executor.go:
```diff
@@ -387,6 +391,82 @@ func (oe *operationExecutor) DetachVolume(
+	actualStateOfWorld ActualStateOfWorldAttacherUpdater) {
+
+	// A map of plugin names and nodes on which they exist with volumes they manage
+	bulkVerifyPluginsByNode := make(map[string]map[types.NodeName][]*volume.Spec)
```
I am thinking this map is actually a map of nodes by plugin?

jingxu97 commented on pkg/volume/util/operationexecutor/operation_executor.go:
```diff
@@ -387,8 +393,84 @@ func (oe *operationExecutor) DetachVolume(
+			if nodeError != nil {
+				glog.Errorf("BulkVerifyVolumes.VerifyVolumesAreAttached verifying volumes on node %q with %v", node, nodeError)
+			}
+			break
```
This function is much larger than other functions in this file. The logic is a little bit tricky, likethe "break" is very critical in line 449. Could we change something like following?
```
bulkVerifyVolumesByPlugin := make(map[string]map[types.NodeName][]AttachedVolume)
nonBulkVerifyVolumesByPlugin := make(map[string]map[types.NodeName][]AttachedVolume)
for node, nodeAttachedVolumes := range attachedVolumes {
 		for _, volumeAttached := range nodeAttachedVolumes {
 			// get volumePlugin
                        // If plugin support bulkVolume, put in bulkVerifyVolumesByPlugin, 
                        // otherwise put in nonBulkVerifyVolumesByPlugin

 for pluginName, bulkVerifyVolumes := range bulkVerifyVolumesByPlugin
             // oe.operationtor.GenerateBulkVolumeBerifyFunc( bulkVerifyVolumes, actualStateOfWorld)

for pluginName, nonBulkVerifyVolumes := range nonBulkVerifyVolumesByPlugin
           for nodeName, volumes := range nonBulkVerifyVolumes 
           //oe.VerifyVolumesAttached(volumes, nodeName, actualStateOfWorld)
```
This might involves more looping, but I feel the logic is easier and code change is smaller.

jingxu97 replied:
talked offline, can leave it as it is now.

jingxu97 commented:
/lgtm

jingxu97 commented:
/lgtm

justinsb commented:
/approve

saad-ali commented:
@k8s-bot test this

gnufied commented:
@k8s-bot kops aws e2e test this

gnufied commented:
@jingxu97 @saad-ali the lgtm got canelled by the rebase. ptal

justinsb commented:
Reapplying LGTM after rebase

/lgtm

@saad-ali can you approve?

justinsb commented:
@derekwaynecarr also I think needs an approve from you for the function rename - or @gnufied you could undo that rename?

jingxu97 commented on pkg/volume/util/operationexecutor/operation_generator.go:
```diff
@@ -167,6 +175,66 @@ func (og *operationGenerator) GenerateVolumesAreAttachedFunc(
+
+				if !nodeChecked {
+					actualStateOfWorld.MarkVolumeAsDetached(volumeSpecMap[volumeSpec], nodeName)
+					continue
```
Put a level 1 or 2 log here, since this is an important operation

gnufied replied:
ack

jingxu97 commented on pkg/volume/util/operationexecutor/operation_generator.go:
```diff
@@ -167,6 +175,66 @@ func (og *operationGenerator) GenerateVolumesAreAttachedFunc(
+					continue
+				}
+
+				check := nodeVolumeSpecs[volumeSpec]
```
what is the purpose of this check? The old VerifyVolumesAreAttached does not have this?

gnufied replied:
The first check:

```nodeVolumeSpecs, nodeChecked := attached[nodeName]```

Checks if node is there in `attached` map, if it isn't - we mark the volume as detached. We are doing this mostly because we don't rely on the fact that  each node will be present in `attached` map. 

The second check is for - if node is present but volume is marked as detached. It is needed because `attached` map is now of type - `map[type.NodeName][*volume.Spec]bool` .

jingxu97 replied:
ok, actually based on the implementation of BulkVerifyVolumes, there is no need to check nodeChecked because it must in the attached map.

gnufied replied:
Yeah, but it is only later I changed returned value in `BulkVerifyVolumes` to always return an full initialized map. 

But anyhow, I believe we shouldn't rely on node keys to be always present. maps are by their nature opaque. So, if we remove node based check, we would start throwing unhandled panic errors if *somehow* the returned map isn't fully initialized (and by that, I mean every node is set etc). 

basically, it is better to be safe than sorry. :-)

jingxu97 replied:
but if the node is missing from the map, is it safe to mark volume as detached? Or should we assume we don't know anything about the node status and should not change it?

gnufied replied:
okay I have dropped marking volume as detached if that happens, just log and continue.

saad-ali commented:
/lgtm
/approve

saad-ali commented:
/lgtm
/approve

gnufied commented:
@justinsb I have removed naming change but that still means there will be a change in admission_test.go because the interface of `DisksAreAttached` function has changed.

I will try and get hold of @derekwaynecarr and get lgtm from him tomorrow.

gnufied commented:
@k8s-bot gci gke e2e test this

gnufied commented:
@k8s-bot cvm gke e2e test this

jingxu97 commented on pkg/cloudprovider/providers/aws/aws.go:
```diff
@@ -330,8 +330,7 @@ type Volumes interface {
 
-	// Check if a list of volumes are attached to the node with the specified NodeName
-	DisksAreAttached(diskNames []KubernetesVolumeID, nodeName types.NodeName) (map[KubernetesVolumeID]bool, error)
+	DisksAreAttached(map[types.NodeName][]KubernetesVolumeID) (map[types.NodeName]map[KubernetesVolumeID]bool, error)
```
add comments like other functions?

gnufied replied:
fixed.

jingxu97 commented on pkg/cloudprovider/providers/aws/aws.go:
```diff
@@ -1761,36 +1760,64 @@ func (c *Cloud) DiskIsAttached(diskName KubernetesVolumeID, nodeName types.NodeN
+	}
+
+	if len(awsInstances) == 0 {
+		glog.V(3).Infof("No instances found with volumes still attached to node")
```
Maybe make it V(2),  add something like "DisksAreAttached will assume all disks are not attached"

gnufied replied:
fixed.

jingxu97 commented on pkg/cloudprovider/providers/aws/aws.go:
```diff
@@ -330,8 +330,8 @@ type Volumes interface {
 
-	// Check if a list of volumes are attached to the node with the specified NodeName
-	DisksAreAttached(diskNames []KubernetesVolumeID, nodeName types.NodeName) (map[KubernetesVolumeID]bool, error)
+	// Check if disks specified in argument map is still attached to their respective nodes.
```
is --> are

gnufied replied:
fixed

jingxu97 commented on pkg/cloudprovider/providers/aws/aws.go:
```diff
@@ -1761,36 +1761,64 @@ func (c *Cloud) DiskIsAttached(diskName KubernetesVolumeID, nodeName types.NodeN
+	awsInstances, err := c.getInstancesByNodeNames(dnsNameSlice)
 	if err != nil {
-		if err == cloudprovider.InstanceNotFound {
+		return attached, err
```
I am ok with this. But wondered whether it better to return nil if there is an error?

jingxu97 commented:
/lgtm

gnufied commented:
@k8s-bot non-cri node e2e test this

gnufied commented:
@k8s-bot gce etcd3 e2e test this

derekwaynecarr commented:
/approve

ethernetdan commented:
@k8s-bot gci gke e2e test this
@k8s-bot kubemark e2e test this
@k8s-bot cvm gke e2e test this
@k8s-bot cvm gce e2e test this
@k8s-bot non-cri e2e test this
@k8s-bot gce etcd3 e2e test this
@k8s-bot gci gce e2e test this

gnufied commented:
What does this do on a merged request? As far as I could see, the failures in tests were because of unrelated authorization problems and such.

BaluDontu commented:
I am working on introducing BulkVerifyVolumes() API in vsphere similar to one you introduced in AWS at https://github.com/kubernetes/kubernetes/pull/41306.

Here I see BulkVerifyVolumes API returns map[types.NodeName]map[*volume.Spec]bool and error. For vsphere, I am trying to implement the functionality and see the following roadblocks.

1. In vsphere, when a node is powered off the volumes are not detached automatically as opposed to case in GCE and AWS. I see the existing BulkVerifyVolumes API throws an error for the entire API call but not for individual nodes. In case consider that I have 2 nodes with 2 volumes attached to each node. If node2 is down, the BulkVerifyVolumes should throw an error for that particular nodeVM which is node2, and then only kubernetes attempts to detach the disk for that node. For node 1, it returns vol1 - true and vol2 - true which shows that the volumes are still attached. With the existing API, I cannot return an error for the individual node, I have to throw an error for the entire API which doesn't seem to be helpful in this because its only node2 which is down and needs volumes to be detached automatically and not node1. 

Overall with existing API, we cannot send the status for every node rather we can send it only for the entire API call. This might seem perfect in envt. like AWS and GCE which volumes are detached automatically when the node is powered off. 

@jingxu97 @gnufied Can you please provide your inputs on this.

gnufied commented:
@BaluDontu you can return error state just for Node 2.  But it depends on how batch fetching of volume state works in Vsphere.  Using AWS as an example - https://github.com/kubernetes/kubernetes/blob/master/pkg/volume/aws_ebs/attacher.go#L126 can you not just return `false` for Node2 and each volume spec attached to that node?
